src/System.Private.CoreLib/shared/System/BitOps.cs

```diff
@@ -49,15 +57,183 @@ public static int TrailingZeroCount(uint value)
+        /// </summary>
+        /// <param name="value">The value.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static int LeadingZeroCount(uint value)
+        {
+            if (Lzcnt.IsSupported)
+            {
+                // Note that LZCNT contract specifies 0->32
+                return (int)Lzcnt.LeadingZeroCount(value);
+            }
+
+            // Software fallback has behavior 0->0, so special-case to match intrinsic path 0->32
+            if (value == 0)
+            {
+                return 32;
+            }
+
+            return 31 - Log2(value);
+        }
+
+        /// <summary>
+        /// Count the number of leading zero bits in a mask.
+        /// Similar in behavior to the x86 instruction LZCNT.
+        /// </summary>
+        /// <param name="value">The value.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static int LeadingZeroCount(ulong value)
+        {
+            if (Lzcnt.X64.IsSupported)
+            {
+                // Note that LZCNT contract specifies 0->64
+                return (int)Lzcnt.X64.LeadingZeroCount(value);
+            }
+
+            uint hi = (uint)(value >> 32);
+
+            if (hi == 0)
+            {
+                return 32 + LeadingZeroCount((uint)value);
+            }
+
+            return LeadingZeroCount(hi);
+        }
+
+        /// <summary>
+        /// Returns the integer (floor) log of the specified value, base 2.
+        /// Note that by convention, input value 0 returns 0 since Log(0) is undefined.
+        /// </summary>
+        /// <param name="value">The value.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static int Log2(uint value)
+        {
+            // value    lzcnt   actual  expected
+            // ..0000   32      0        0 (by convention, guard clause)
+            // ..0001   31      31-31    0
+            // ..0010   30      31-30    1
+            // 0010..    2      31-2    29
+            // 0100..    1      31-1    30
+            // 1000..    0      31-0    31
+            if (Lzcnt.IsSupported)
+            {
+                // Enforce conventional contract 0->0 (since Log(0) is undefined)
+                if (value == 0)
+                {
+                    return 0;
+                }
+
+                // Note that LZCNT contract specifies 0->32
+                return 31 - (int)Lzcnt.LeadingZeroCount(value);
+            }
+
+            // Already has contract 0->0, without branching
+            return Log2SoftwareFallback(value);
+        }
+
+        /// <summary>
+        /// Returns the integer (floor) log of the specified value, base 2.
+        /// Note that by convention, input value 0 returns 0 since Log(0) is undefined.
+        /// Does not incur branching.
+        /// </summary>
+        /// <param name="value">The value.</param>
+        private static int Log2SoftwareFallback(uint value)
```

Comment:
nit: This could probably just be a local function. Once the PR supporting C# 8 goes in, we can even shadow the parameter name which makes it even "cleaner".

Reply:
@tannergooding do you mind if I look at this in the next PR which is waiting to go, I would like to get this one in?

Reply:
It may be better if this stays as is, and instead `LeadingZeroCount` is changed to call `Log2SoftwareFallback` directly, without going through extra `Log2` layer.

Reply:
OK, removed extra `Log2` layer in subsequent PR